pkg/didcomm/protocol/didexchange/persistence.go

```diff
@@ -94,13 +104,87 @@ func (c *ConnectionRecorder) GetInvitation(id string) (*Invitation, error) {
 	return result, nil
 }
 
-// GetConnectionRecord return connection record
+// GetConnectionRecord return connection record based on the connection ID
 func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
-	name, err := c.store.Get(connectionID)
+	k := connectionKeyPrefix(connectionID)
+	connRecordBytes, err := c.store.Get(k)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("get connection record:%w", err)
+	}
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// GetConnectionRecordByNSThreadID return connection record via name spaced threadID
+func (c *ConnectionRecorder) GetConnectionRecordByNSThreadID(nsThreadID string) (*ConnectionRecord, error) {
+	connectionIDBytes, err := c.store.Get(nsThreadID)
+	if err != nil {
+		return nil, fmt.Errorf("get connectionID by name spaced threadID :%w", err)
+	}
+	// adding prefix for storing connection record
+	k := connectionKeyPrefix(string(connectionIDBytes))
+
+	connRecordBytes, err := c.store.Get(k)
+	if err != nil {
+		return nil, fmt.Errorf("get connection record by connectionID:%w", err)
+	}
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// saveConnectionRecord saves the connection record against the connection id  in the store
+func (c *ConnectionRecorder) saveConnectionRecord(record *ConnectionRecord) error {
+	k := connectionKeyPrefix(record.ConnectionID)
+	bytes, err := json.Marshal(record)
+	if err != nil {
+		return fmt.Errorf("save connection record:%w", err)
 	}
-	return &ConnectionRecord{State: string(name)}, nil
+	return c.store.Put(k, bytes)
+}
+
+// saveNewConnectionRecord saves newly created connection record against the connection id in the store
+// and it creates mapping from name spaced ThreadID to connection ID
+func (c *ConnectionRecorder) saveNewConnectionRecord(record *ConnectionRecord) error {
+	err := record.isValid()
+	if err != nil {
+		return err
+	}
+	err = c.saveConnectionRecord(record)
+	if err != nil {
+		return fmt.Errorf("save new connection record:%w", err)
+	}
+	return c.saveNSThreadID(record.ThreadID, record.Namespace, record.ConnectionID)
+}
+
+func (c *ConnectionRecorder) saveNSThreadID(thid, namespace, connectionID string) error {
+	if namespace != myNSPrefix && namespace != theirNSPrefix {
+		return fmt.Errorf("namespace not supported")
+	}
+
+	prefix := myNSPrefix
+	if namespace == theirNSPrefix {
+		prefix = theirNSPrefix
+	}
+
+	k, err := createNSKey(prefix, thid)
+	if err != nil {
+		return err
+	}
+
+	return c.store.Put(k, []byte(connectionID))
+}
+func prepareConnectionRecord(connRecBytes []byte) (*ConnectionRecord, error) {
+	connRecord := &ConnectionRecord{}
+	err := json.Unmarshal(connRecBytes, connRecord)
+	if err != nil {
+		return nil, fmt.Errorf("prepare connection record: %w", err)
+	}
+	return connRecord, nil
+}
+func (r *ConnectionRecord) isValid() error {
```

Comment:
It would be nice if we can group this method with ConnectionRecord definition (since both are in same file).

Reply:
done